File: app/src/main/java/android/mahendra/attendancemanager/repositories/PeriodRepository.java
package android.mahendra.attendancemanager.repositories;

import android.app.Application;
import android.mahendra.attendancemanager.daos.PeriodDao;
import android.mahendra.attendancemanager.daos.SubjectPeriodDao;
import android.mahendra.attendancemanager.database.MainDatabase;
import android.mahendra.attendancemanager.models.Period;
import android.mahendra.attendancemanager.models.Subject;

import androidx.lifecycle.LiveData;

import java.util.List;

public class PeriodRepository {
    private PeriodDao mPeriodDao;
    private SubjectPeriodDao mSubjectPeriodDao;
    private LiveData<List<Period>> mAllPeriods;

    public PeriodRepository(Application application) {
        MainDatabase db = MainDatabase.getDatabase(application);
        mPeriodDao = db.mPeriodDao();
        mSubjectPeriodDao = db.mSubjectPeriodDao();
        mAllPeriods = mPeriodDao.getAllPeriods();
    }

    public void insert(Period period) {
        MainDatabase.databaseWriteExecutor.execute(() -> {
            mPeriodDao.insert(period);
        });
    }

    public void update(Period period) {
        MainDatabase.databaseWriteExecutor.execute(() -> {
            mPeriodDao.update(period);
        });
    }

    public void delete(Period period) {
        MainDatabase.databaseWriteExecutor.execute(() -> {
            mPeriodDao.delete(period);
        });
    }

    public LiveData<List<Period>> getAllPeriods() {
        return mAllPeriods;
    }

    public LiveData<List<Period>> getAllPeriodsOn(int weekDay) {
        return mPeriodDao.getAllPeriodsOn(weekDay);
    }

    public LiveData<List<Period>> getAllPeriodsOfSubject(String subjectTitle) {
        return mPeriodDao.getAllPeriodsOfSubject(subjectTitle);
    }

    public LiveData<List<Subject>> getAllSubjectsOn(int weekDay) {
        return mSubjectPeriodDao.getAllSubjectsOn(weekDay);
    }

    public void deletePeriod(int periodNumber, int weekDay) {
        MainDatabase.databaseWriteExecutor.execute(() ->{
                mPeriodDao.deletePeriod(periodNumber, weekDay);
        });
    }
}

File: app/src/main/java/android/mahendra/attendancemanager/repositories/PeriodRepository.kt
package android.mahendra.attendancemanager.repositories

import android.app.Application
import android.mahendra.attendancemanager.daos.PeriodDao
import android.mahendra.attendancemanager.daos.SubjectDao
import android.mahendra.attendancemanager.daos.SubjectPeriodDao
import android.mahendra.attendancemanager.database.MainDatabase
import android.mahendra.attendancemanager.models.Period
import android.mahendra.attendancemanager.models.Subject

import androidx.lifecycle.LiveData

class PeriodRepository private constructor(
        private val periodDao: PeriodDao,
        private val subjectPeriodDao: SubjectPeriodDao
) {
    val allPeriods: LiveData<List<Period>> = periodDao.getAllPeriods()

    suspend fun insert(period: Period) {
            periodDao.insert(period)
    }

    suspend fun update(period: Period) {
            periodDao.update(period)
    }

    suspend fun delete(period: Period) {
            periodDao.delete(period)
    }

    fun getAllPeriodsOn(weekDay: Int): LiveData<List<Period>> {
        return periodDao.getAllPeriodsOn(weekDay)
    }

    fun getAllPeriodsOfSubject(subjectTitle: String): LiveData<List<Period>> {
        return periodDao.getAllPeriodsOfSubject(subjectTitle)
    }

    fun getAllSubjectsOn(weekDay: Int): LiveData<List<Subject>> {
        return subjectPeriodDao.getAllSubjectsOn(weekDay)
    }

    fun getPeriod(periodId: Long): LiveData<Period> {
        return periodDao.getPeriod(periodId)
    }

    suspend fun deletePeriod(periodNumber: Int, weekDay: Int) {
        periodDao.deletePeriod(periodNumber, weekDay)
    }

    companion object {

        @Volatile private var instance: PeriodRepository? = null

        fun getInstance(periodDao: PeriodDao, subjectPeriodDao: SubjectPeriodDao) =
                instance ?: synchronized(this) {
                    instance ?: PeriodRepository(periodDao, subjectPeriodDao).also { instance = it }
                }
    }
}
